renderers/termbox2/clay_renderer_termbox2.c

```diff
@@ -0,0 +1,1190 @@
+                                } else if ((x < border_skip_begin_x)
+                                    && (y >= border_skip_end_y)) { // Bottom left
+                                    ch = U'\u2514';
+                                } else if ((x >= border_skip_end_x)
+                                    && (y >= border_skip_end_y)) { // Bottom right
+                                    ch = U'\u2518';
+                                } else if (x < border_skip_begin_x || x >= border_skip_end_x) {
+                                    ch = U'\u2502';
+                                } else if (y < border_skip_begin_y || y >= border_skip_end_y) {
+                                    ch = U'\u2500';
+                                }
+                                break;
+                            }
+                            case CLAY_TB_BORDER_CHARS_DEFAULT:
+                            case CLAY_TB_BORDER_CHARS_ASCII: {
+                                if ((x < border_skip_begin_x || x >= border_skip_end_x)
+                                    && (y < border_skip_begin_y || y >= border_skip_end_y)) {
+                                    ch = '+';
+                                } else if (x < border_skip_begin_x || x >= border_skip_end_x) {
+                                    ch = '|';
+                                } else if (y < border_skip_begin_y || y >= border_skip_end_y) {
+                                    ch = '-';
+                                }
+                                break;
+                            }
+                        }
+
+                        color_pair color_bg_new = clay_tb_get_transparency_color(
+                            x, y, (color_pair) { color_bg, color_tb_bg });
+                        clay_tb_set_cell(
+                            x, y, ch, color_tb_fg, color_bg_new.termbox, color_bg_new.clay);
+                    }
+                }
+                break;
+            }
+            case CLAY_RENDER_COMMAND_TYPE_TEXT: {
+                Clay_TextRenderData render_data = command->renderData.text;
+                Clay_Color color_fg = render_data.textColor;
+                uintattr_t color_tb_fg = clay_tb_color_convert(color_fg);
+
+                Clay_StringSlice *text = &render_data.stringContents;
+                int32_t i = 0;
+                for (int y = box_begin_y; y < box_end_y; ++y) {
+                    for (int x = box_begin_x; x < box_end_x;) {
+                        uint32_t ch = ' ';
+                        if (i < text->length) {
+                            int codepoint_length = tb_utf8_char_to_unicode(&ch, text->chars + i);
+                            if (0 > codepoint_length) {
+                                clay_tb_assert(false, "Invalid utf8");
+                            }
+                            i += codepoint_length;
+
+                            uintattr_t color_tb_bg = clay_tb_color_buffer_termbox_get(x, y);
```

Comment:
That's awesome! I've been working on a renderer too for tembox2 but at this point yours looks more complete. 

I know that termbox2 buffer is a private API and you implemented a bespoke color buffer to access background colors for text. I took the approach of adding background color to borders and text instead with the idea that conceptually nothing stops TEXT and BORDER to have a background color. It feels simpler and more explicit when defining text in a Clay layout but is kind of redundant too. Here was the question I asked: https://github.com/nicbarker/clay/issues/409

What are your thoughts on both approaches?

Reply:
That makes perfect sense. I've overlooked text overflows 👍